### PR TITLE
Refactor the CAS handler

### PR DESCRIPTION
**Pull request by clokep**
This is part of #8856 reordered and refactored a bit. Each commit is reviewable separately, but the overview is:

* Return a `CasResponse` object instead of a tuple (this is just a convenience).
* Render errors to the user instead of raising exceptions.
* Refactor some of the methods.

### CODE REVIEW

**Review by richvdh — CHANGES_REQUESTED** [review]
generally lgtm but I'm worried about the HTTP response codes changing.

File: synapse/handlers/cas_handler.py [hunk 1]
```diff
@@ -229,19 +231,32 @@ async def handle_ticket(
...
+            self._sso_handler.render_error(request, e.error, e.error_description)
+            return
 
         for required_attribute, required_value in self._cas_required_attributes.items():
             # If required attribute was not in CAS Response - Forbidden
             if required_attribute not in cas_response.attributes:
-                raise LoginError(401, "Unauthorized", errcode=Codes.UNAUTHORIZED)
+                self._sso_handler.render_error(
+                    request, "unauthorised", "You are not authorised to log in here."
+                )
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/cas_handler.py:277]: I'm not sure why this matters sine it is showing the user an error page (via an iframe or such), not responding to an API. This matches what we do for SAML/OIDC unless I'm missing something about how CAS works that makes it different.

[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/handlers/cas_handler.py:277]: well, if nothing else, it helps distinguish the cases in access logs and metrics.

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/cas_handler.py]: Fair enough. The codes should be consistent with before the refactor now.

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**